vxsandbox/rlimiter.py

```diff
@@ -1,102 +1,78 @@
-        self._args = [self._executable] + argv[start + 1:]
-        self._env = env
-
-    def _apply_rlimits(self):
-        data = os.environ[self._SANDBOX_RLIMITS_]
-        rlimits = json.loads(data) if data.strip() else {}
-        for rlimit, (soft, hard) in rlimits.iteritems():
-            # Cap our rlimits to the maximum allowed.
-            rsoft, rhard = resource.getrlimit(int(rlimit))
-            soft = min(soft, rsoft)
-            hard = min(hard, rhard)
-            resource.setrlimit(int(rlimit), (soft, hard))
-
-    def _reset_signals(self):
-        # reset all signal handlers to their defaults
-        for i in range(1, signal.NSIG):
-            if signal.getsignal(i) == signal.SIG_IGN:
-                signal.signal(i, signal.SIG_DFL)
-
-    def _sanitize_fds(self):
-        # close everything except stdin, stdout and stderr
-        maxfds = resource.getrlimit(resource.RLIMIT_NOFILE)[1]
-        os.closerange(3, maxfds)
-
-    def execute(self):
-        self._apply_rlimits()
-        self._restore_child_env(os.environ)
-        self._sanitize_fds()
-        self._reset_signals()
-        os.execvpe(self._executable, self._args, self._env)
-
-    _SANDBOX_RLIMITS_ = "_SANDBOX_RLIMITS_"
 
-    @classmethod
-    def _override_child_env(cls, env, rlimits):
-        """Put RLIMIT config in the env."""
-
-        env[cls._SANDBOX_RLIMITS_] = json.dumps(rlimits)
+    ULIMIT_PARAMS = {
+        resource.RLIMIT_CORE: "c",
+        resource.RLIMIT_CPU: "t",
+        resource.RLIMIT_FSIZE: "f",
+        resource.RLIMIT_DATA: "d",
+        resource.RLIMIT_STACK: "s",
+        resource.RLIMIT_RSS: "m",
+        resource.RLIMIT_NOFILE: "n",
+        resource.RLIMIT_MEMLOCK: "l",
+        resource.RLIMIT_AS: "v",
+    }
+
+    # These ones can't always be set, so ignore them.
+    ULIMIT_SKIP = (
+        resource.RLIMIT_STACK,
+        resource.RLIMIT_MEMLOCK,
+    )
+
+    def __init__(self, rlimits, args, **kwargs):
+        self._args = args
+        self._rlimits = rlimits
+        self._kwargs = kwargs
+
+    def execute(self, reactor, protocol):
+        reactor.spawnProcess(
+            protocol, '/bin/bash', args=self.build_args(), **self._kwargs)
+
+    def build_args(self):
+        return ['bash', '-e', '-c', self.build_script(), '--'] + self._args
+
+    def build_script(self):
+        script_lines = ["#!/bin/bash", ""]
+        script_lines.extend(self._build_rlimit_commands())
+        script_lines.extend(["", 'exec "$@"'])
+        return "\n".join(script_lines)
+
+    def _build_rlimit_commands(self):
+        yield "# Set resource limits."
+        for rlimit, (soft, hard) in sorted(self._rlimits.items()):
+            if rlimit in self.ULIMIT_SKIP:
+                # We can't set this one.
+                continue
+            param = self.ULIMIT_PARAMS[rlimit]
+            rsoft, rhard = resource.getrlimit(int(rlimit))
+            yield "ulimit -S%s %s" % (param, minpositive(soft, rsoft))
```

Comment:
No, because we don't want to raise the soft limit if it's already lower than what we have configured.